examples_tests/41.VisibilityBuffer/main.cpp

```diff
@@ -3,815 +3,437 @@
-        uint32_t bufoffset = 0u;
-        for (uint32_t i = 0u; i < regions->size(); ++i)
-        {
-            auto& region = (*regions)[i];
-            region.bufferImageHeight = 0u;
-            region.bufferOffset = bufoffset;
-            region.bufferRowLength = paddedExtent>>i;
-            region.imageExtent = {paddedExtent>>i,paddedExtent>>i,1u};
-            region.imageOffset = {0u,0u,0u};
-            region.imageSubresource.baseArrayLayer = 0u;
-            region.imageSubresource.layerCount = 1u;
-            region.imageSubresource.mipLevel = i;
-
-            bufoffset += texelBytesize*region.imageExtent.width*region.imageExtent.height;
-        }
-        auto buf = core::make_smart_refctd_ptr<asset::ICPUBuffer>(bufoffset);
-        paddedImg->setBufferAndRegions(std::move(buf), regions);
-    }
+    //setOffsetTables
+
+    core::vector<uint32_t> offsetTableLocal(offsetTableSz);
+
+    for (uint32_t i = 0u; i < offsetTableLocal.size(); i++)
+        offsetTableLocal[i] = cdot[i].attribOffset[0];
+
+    offsetTable[0].offsetBuffer.offset = 0u;
+    offsetTable[0].offsetBuffer.buffer = driver->createFilledDeviceLocalGPUBufferOnDedMem(sizeof(uint32_t) * offsetTableLocal.size(), static_cast<void*>(offsetTableLocal.data()));
+
+    for (uint32_t i = 0u; i < offsetTableLocal.size(); i++)
+        offsetTableLocal[i] = cdot[i].attribOffset[1];
+
+    offsetTable[1].offsetBuffer.offset = 0u;
+    offsetTable[1].offsetBuffer.buffer = driver->createFilledDeviceLocalGPUBufferOnDedMem(sizeof(uint32_t) * offsetTableLocal.size(), static_cast<void*>(offsetTableLocal.data()));
 
-    //copy mip 0 to new image while filling padding according to wrapping modes
-    asset::CPaddedCopyImageFilter::state_type copy;
-    copy.axisWraps[0] = _wrapu;
-    copy.axisWraps[1] = _wrapv;
-    copy.axisWraps[2] = asset::ISampler::ETC_CLAMP_TO_EDGE;
-    copy.borderColor = asset::ISampler::ETBC_FLOAT_OPAQUE_BLACK;
-    copy.extent = params.extent;
-    copy.layerCount = 1u;
-    copy.inMipLevel = 0u;
-    copy.inOffset = {0u,0u,0u};
-    copy.inBaseLayer = 0u;
-    copy.outOffset = {0u,0u,0u};
-    copy.outBaseLayer = 0u;
-    copy.outMipLevel = 0u;
-    copy.paddedExtent = {paddedExtent,paddedExtent,1u};
-    copy.relativeOffset = {0u,0u,0u};
-    copy.inImage = _img;
-    copy.outImage = paddedImg.get();
-
-    asset::CPaddedCopyImageFilter::execute(&copy);
-
-    using mip_gen_filter_t = asset::CMipMapGenerationImageFilter<asset::CBoxImageFilterKernel,asset::CBoxImageFilterKernel>;
-    //generate all mip levels
+    for (uint32_t i = 0u; i < offsetTableLocal.size(); i++)
+        offsetTableLocal[i] = cdot[i].attribOffset[2];
+
+    offsetTable[2].offsetBuffer.offset = 0u;
+    offsetTable[2].offsetBuffer.buffer = driver->createFilledDeviceLocalGPUBufferOnDedMem(sizeof(uint32_t) * offsetTableLocal.size(), static_cast<void*>(offsetTableLocal.data()));
+}
+
+void setPipeline(IVideoDriver* driver, ICPUSpecializedShader* vs, ICPUSpecializedShader* fs,
```

Comment:
call it create not set, you're not binding the pipeline to be bind point inside.